drivers/dac/ad5754r/ad5754r.c

```diff
@@ -0,0 +1,767 @@
+ */
+int ad5754r_update_bits(struct ad5754r_dev *dev,
+			uint8_t instr_addr,
+			uint16_t mask,
+			uint16_t reg_val)
+{
+	int ret;
+	uint16_t reg_data;
+
+	ret = ad5754r_read(dev, instr_addr, &reg_data);
+	if (ret)
+		return ret;
+
+	reg_data &= ~mask;
+	reg_data |= reg_val;
+
+	return ad5754r_write(dev, instr_addr, reg_data);
+}
+
+/**
+ * @brief Update DAC register value for specific channel.
+ * @param dev - The device structure.
+ * @param chn - The index of channel being addressed.
+ * @param value - The data to be written to the dac channel.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad5754r_update_dac_ch_register(struct ad5754r_dev *dev,
+				   uint8_t chn,
+				   uint16_t value)
+{
+	if (chn > AD5754R_NUM_CHANNELS)
+		return -EINVAL;
+
+	return ad5754r_write(dev, AD5754R_PREP_INSTR_ADDR(AD5754R_REG_DAC, chn),
+			     value);
+}
+
+/**
+ * @brief Update DAC register value for all channels.
+ * @param dev - The device structure.
+ * @param value - The data to be written to the dac channel registers.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad5754r_update_dac_all_ch_registers(struct ad5754r_dev *dev,
+					uint16_t value)
+{
+	return ad5754r_write(dev,
+			     AD5754R_PREP_INSTR_ADDR(AD5754R_REG_DAC,
+					     AD5754R_DAC_CH_ALL_ADDR),
+			     value);
+}
+
+/**
+ * @brief Read DAC register value for specific channel.
+ * @param dev - The device structure.
+ * @param chn - The index of channel being addressed.
+ * @param value - The data read from dac channel register.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad5754r_read_dac_ch_register(struct ad5754r_dev *dev,
+				 uint8_t chn,
+				 uint16_t *value)
+{
+	if (!value || (chn >= AD5754R_NUM_CHANNELS))
+		return -EINVAL;
+
+	return ad5754r_read(dev,
+			    AD5754R_PREP_INSTR_ADDR(AD5754R_REG_DAC, chn),
+			    (uint16_t *)value);
+}
+
+/**
+ * @brief Convert DAC code to millivolts for specific channel.
+ * @param dev - The device structure.
+ * @param chn - The index of channel being addressed.
+ * @param code - DAC code
+ * @param vref - Input reference voltage to AD5754R
+ * @param value - Millivolts output calculated using DAC code.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad5754r_dac_code_to_mvolts(struct ad5754r_dev *dev, uint8_t chn,
+			       uint16_t code, int vref,
+			       uint16_t *value)
+{
+	int ret;
+	uint8_t bin_2scomp;
+
+	if (!dev || !value || (chn >= AD5754R_NUM_CHANNELS))
+		return -EINVAL;
+
+	/* Based on unipolar/bipolar output range, calculate milivolts */
+	if (dev->dac_ch_range[chn] < AD5754R_SPAN_M5V_TO_5V) {
+		*value = ad5754r_gain_values[dev->dac_ch_range[chn]]*vref*code /
+			 (1 << AD5754R_MAX_RESOLUTION);
+	} else {
+		if (!dev->gpio_bin_2scomp) {
+			return -EINVAL; // No way to tell output coding
+		}
+
+		ret = no_os_gpio_get_value(dev->gpio_bin_2scomp, &bin_2scomp);
```

Comment:
This behavior doesn't seems right. I think most of the times, this bin/2scomp pin will be hardwired to either VDD or GND without being routed to a pin of the microcontroller. Moreover, there is no reason the encoding will change while the DAC is powered up. The way I suggest to handle this is to give the encoding as a parameter in the `ad5754_init_param` and set the optional gpio based on that.